### DISCUSSION TITLE
Daylength

### DISCUSSION THREAD

**Discussion by andythenorth** [OpenTTD/OpenTTD#8397] [Features]
Daylength is much discussed. 

Multiple patches exist. All are believed to have issues.  Daylength fundamentally changes time scaling across the game in a way that was never anticipated.  It's a design problem, not simply a programming problem.

So this isn't a 'vote here for daylength' discussion, rather it's to try and capture what people want from the game that leads them to the idea of changing daylength.

See also: groundhog year #8384

**Comment by andythenorth** [comment]
### Slow down vehicle progression

Play 1960 trains forever!  

For many players, the game seems to progress quickly.  A new generation of vehicles appears, so we upgrade and tune our networks.  Then just as we finish, boom!  A new generation of vehicles appears. 

On busy or large maps it can be hard to keep up.

Slower progression of vehicles would make a more relaxing games for some players.

This can be solved in multiple ways, of which daylength is one.  Others include using non-real intro dates in newgrf, or ideas like groundhog year #8384

**Reply by JGRennison** [reply]
I for one find the rate of progression of vehicles in the default timescale overly fast.
Having more new vehicle announcements than can be usefully dealt with soon followed by silence seems a bit odd.

As an aside, reaching the "end" years and having no available vehicles at all for a particular cargo as they have all expired is very annoying.
(I added [this](https://github.com/JGRennison/OpenTTD-patches/commit/38d42795c847c44aa9ac5c4f4c073bc695bc2dfb) just because I got hit by this problem.)

A lot of people (users and GRF authors) are seriously into vehicle realism, endless talk about stats, liveries, etc.
I would imagine that a solution based on non-real dates wouldn't go down so well.

**Reply by michicc** [reply]
It should also be noted that many (not all) players that look for this goal still expect that vehicle speed, income, production etc per "real time" to remain the same.

**Reply by Tintinfan** [reply]
> A lot of people (users and GRF authors) are seriously into vehicle realism, endless talk about stats, liveries, etc.
> I would imagine that a solution based on non-real dates wouldn't go down so well.

Can only second this!

**Comment by andythenorth** [comment]
### Building very large networks on very large maps

I don't know if this one is a factor, but do some players just want a slower game so they can build very large networks on very large maps?

**Reply by 2TallTyler** [reply]
This seems related to slowing down vehicle progression, unless you were getting at another reason why large networks don't work well with default game speed.

**Reply by auge8472** [reply]
Building a network in a single area and observing it to monitoring it's function and for enjoyment can cause building a second network in another corner of the map bein finished in another, next era where my plan for the network (vehicles, visual nature) breaks. So **for me** time elapses often to fast. The only way to solve this conflict for me is to use only very small maps (where I wish to use bigger ones).

**Reply by nielsmh** [reply]
I think this has to do with how the financials put together. They don't want a train crossing a 4096 map to take two years for a round trip, but perhaps a month. Having vehicles that take too long per journey on large maps ruins any "profit per year" statistics, when vehicles may go a year or more without any income.

**Comment by andythenorth** [comment]
### Reduced rate of town and industry cargo generation

I don't know if this one is a factor, but typical approaches to daylength will cause towns and industry to generate less cargo per month, which some players might find preferable?

Based on https://www.tt-forums.net/viewtopic.php?p=1191037#p1191037

This could alternatively be addressed in newgrf.

**Reply by JGRennison** [reply]
Scaling town cargo generation to suit player preferences is quite easy and there are patches to do that.

Scaling industry cargo generation is another matter due to the issue of industry NewGRFs and the industry spec.
As far as I am aware there are no implementations of this.
I presume that this could be done within the NewGRF itself to a limited extent but I'm not aware of any that actually support that.

**Comment by andythenorth** [comment]
### Making realistic simulations of real world transport networks

This one is based on the comment at https://www.tt-forums.net/viewtopic.php?p=1190831#p1190831

**Reply by nielsmh** [reply]
This is basically attempting to tune the time and distance scales to match some specific, real-ish scenario, I guess?

**Comment by JGRennison** [comment]
> Daylength fundamentally changes time scaling across the game in a way that was never anticipated. It's a design problem, not simply a programming problem.

A bit of a general comment, but the scaling of the original game is arbitrary and not in any way self-consistent.
Picking a different point in the space of arbitrary parameters, which wasn't previously anticipated, is not necessarily a bad thing.

**Comment by Rymdkejsaren1** [comment]
Here's what made me look into changing the daylength: we sometimes play an online game with friends on a persistent server, but people can't usually play at the same time; they log in during work, after, or someone may not even have time to play every day. For that purpose, the game progresses much too quickly. Come back after you've missed a day and you're hopelessly left behind. Even when there's room for everyone and we make sure not to hog all the good spots, you'll be missing out.

So I figure it's about the rate of **advancement of technology**, **income**, and **growth rate of towns**. We just want a slower general pace, with the years progressing considerably slower so that we can log in, build a line, and not come back a day later to massive cities and someone has made an absolute killing on their two oil lines that now have been running for decades. It's not about competitiveness as much as having a bit of fun together, so "balance" in that sense isn't the issue.

I have no clue about the capacities of the engine, code, or other problems. I just wanted to chime in to say what my friends and I are looking for.

**Reply by Rymdkejsaren1** [reply]
Of course, if you slow it down a lot and the trains move super slow, that would be quite a drag. So I suppose that could become a problem. You'd still want to see your lines in action and not moving in slow motion.

**Reply by kaomoneus** [reply]
> Here's what made me look into changing the daylength: we sometimes play an online game with friends on a persistent server, but people can't usually play at the same time

I 120% agree with @Rymdkejsaren1 ! This is exactly I was wanted game pace changing for. I just want same game balance, but I want it to be just slooower. And yet keeping animation smooth.

I wanted to log into server, spend there 1-2 hours after work, and leave it for next day. Or same but on weekly basis.

Good example is a [Merge Dragons game](https://mergedragons.com/), my wife's favorite. And I kind of envy, haha! She grows dragons, sends them to do some tasks. Then dragons get tired and she had to wait until next day.

AFAIK in OpenTTD "time" is represented by different game mechs which acts a bit independently. And the problem is to gather all of them to scale simultaneously.

Anybody knows what is the most recent patch? Is here a full list of mechs which we should alter for such feature (town growth, cargo manufacturing and so on)?

Thanks!

**Comment by nielsmh** [comment]
My preferred approach to "daylength" would be separating the "world history calendar" from the "game economy calendar".

A while ago I tried starting a patch that separated these such that you could scale up/down the speed of the "world history calendar" (mainly affecting vehicle types, industry types, and town house types available), while the game economy was changed to run on something resembling the player's wall clock time.
I changed the tick length (100% game speed) so 74 ticks/one day translated to very precisely 2 seconds, meaning one average 30 day month would be 1 minute on the clock. Then went about changing all the economy logic so things like industry production was calculated in "per minute" (i.e. changing "game economy months" to all be 30 days), and budget happening in 12 minute (360 day) increments.

The result would become having two clocks going on. One major issue I bumped into was actually naming things. Those new "economy years" (12 minute increments) need a snappy name that's still easy to understand, but also won't get confused with anything in the "world history calendar" time.
I don't remember exactly where I stopped/gave up, but I don't think it got to a playable state.

**Reply by nielsmh** [reply]
One obvious issue with this approach is that is fundamentally changes the presentation of the game, and even at a default pace affects the game economy in small ways, since years become 5 or 6 days shorter.

**Reply by kaomoneus** [reply]
I would enhance, generalize or rephrase this notion to following. 

* **A** There is a physical progress. It is associated with what we call "time" from one side. And from another site it is associated with some achievements (revenue, towns growth, etc).
   
   E.g. I have a bus, and it makes $500 a month. So I watch game calendar and see 1st Jan and $0, and then "few moments later" I check it again and observe 1st Feb and $500.
* **B** There is an indication of physical progress. And here it goes. Speaking about bus, it could be displayed with many ways:
   * as a single trip (fast game pace) with $600 gross - $100 of running cost
   * as like 100 of trips (slow pace) with same total financial indicators.

And here outcomes two things:
1. Game balance is configured as a set of rules for physical progress (*A*)
2. User experience depends on both *A* and *B* + desirable "game pace".

Also another important thing:
* Cargo unit and metric unit (space cell) are merely ways to indicate physical progress. Those may vary when we adjusting game pace.
   * For fast game pace bus transfers only 40 passengers a month with total distance of 30 cells.
   * For slow game pace bus transfers like 4k passengers with total distance of 3k cells.

So as an idea I propose to
* collect rules of *physical progress* using original game settings
* alter *daylength*
* Adjust fares and perhaps towns & cities growth rates
* profit

**Reply by nielsmh** [reply]
@kaomoneus I think that would quickly result in very difficult to understand in-game UI.

But let's run with it. Take my suggestion (since that's what you're replying to) with the separation of calendar and economy. The calendar runs as it does, with days of whatever tick length the player has chosen (default=74).

Then your suggestion would be to have a gigantic lump of additional settings like this?
- Industries produce cargo every N seconds (default 7, which is actually a rounded display of 256 ticks)
- Vehicles pay running cost every N seconds (default 60, which is 30 economy days)
- Lifetime of vehicles is N minutes per designed lifetime in years (default 12)
- Snowy and desert towns require resource deliveries every N seconds for growth (default 60, and citybuilder game scripts need to be aware of this)
And a lot more similar to this, I can't list out right now. This all becomes a UI nightmare, which I personally don't think has a place in "vanilla". And all of this could just as well be configured by NewGRF. Pretty much any cost- or production-scaling you can think of is already available as NewGRF.

My opinion is that the best solution is to allow the core game to separate the calendar from the economy time, and let any cost- and production-scaling the player wants be handled by NewGRF and GS. The separation of calendar from economy time is the only thing that isn't possible to handle gracefully in mods, so only that belongs in the core game.

**Reply by kaomoneus** [reply]
Well, I think I'm a bit out of what you mean by "world history calendar" and "game economy calendar". Could you please explain on some simple example?
Thanks!

**Reply by nielsmh** [reply]
World history calendar:
- Introduction and obsolescence of vehicle models
- Introduction and obsolescence of industry types
- Introduction and obsolescence of town building types
- Introduction and obsolescence of rail, road, and bridge types
By changing the number of ticks a world history day takes you can speed up and slow down the pace of technological change in the world. This affects how much time you can spend in each "era" of the game.

Economy calendar:
- How often industries produce cargo
- How often running costs are paid
- How often loan interest is paid
- How often industries change production rate and how long they can stay without service before shutting down
- How fast the reliability of vehicles deteriorate
- How long vehicles can stay in service before they need replacement
- How fast towns grow
- How often towns in desert and snow regions need deliveries of special cargo to be able to grow
- Many more things...
All of these affect the economical balance of the gameplay, such as how many vehicles you need and the optimal length of routes. Some are roughly equivalent to scaling costs, while others don't have an obvious cost scaling.

My suggestion is decoupling those two concepts such that the economy calender runs on "real time" where everything is measured in minutes and seconds, instead of months and days.
The items in the world history calendar list can't really be changed by modding, without making the year scale look wrong. For that reason, the scale of those things should be changed by the core game, by simply making days longer or shorter (more or fewer ticks per world calendar day).
The items in the economy calendar list can all to some degree be changed by mods and difficulty settings. They don't need any more core game support.

**Comment by kaomoneus** [comment]
Hi all! I completed my work on *Slow Pace mode*. I tried to grab best from all proposals.

[![Video: OpenTTD Slow Pace](https://img.youtube.com/I8bn7Gb7gu8/0.jpg)](https://www.youtube.com/watch?v=I8bn7Gb7gu8)

Detailed description: [SlowPace.md](https://github.com/kaomoneus/OpenTTD/blob/slow_pace/SlowPace.md)

Link to modified game clients: [Latest OpenTTD Slow Pace release](https://github.com/kaomoneus/OpenTTD/releases/latest)

As a short answer to @nielsmh , most of mechs are in fact "animation bound". For example, cargo growth rate. This is bound to vehicle trips (which is animation). So you load vehicle every N animation time units, and thus cargo should also grow with animation pace.

Whatever handled by `OnNewDay` and `OnNewMonth` is handled fine. Whatever handled by `Tick` but meant some non-animation event required to be fixed.

**Reply by 2TallTyler** [reply]
This is very interesting. I appreciate how you've explained which events belong in which of the two time streams, and how you've balanced the game economy to respect the new time scale.

The way you choose the pace in the new game window was initially confusing to me before I wrapped my head around it, but makes sense now. I wonder if there's a need for players to select a custom pace, as there's quite a range between your preset options.

**Reply by James103** [reply]
> I wonder if there's a need for players to select a custom pace, as there's quite a range between your preset options.

I think that there should be a way for players to select a custom pace. For example, some people may think that even 1 hour per year is too fast, but that 1 day per year feels too slow, while some servers may need the game to run at 24 minutes per in-game day in order to make each minute exactly one second long.

---

From the attached file:
> First of all `DAY_TICKS` became a variable which is set when you start new game. So user can select how fast the game pace should be.

The variable could be exposed directly as a ~~"ticks per day" or~~ "real-life seconds per day" value, keeping in mind that the game runs at 1000/30 &#x2248; 33.33 ticks per real life second. For example, `DAY_TICKS` should be 100/3 * 60 * 24 = 48000 in order to get 24 minutes per day.

**Reply by 2TallTyler** [reply]
I would argue against "ticks per day" being exposed to the player, as an unnecessarily technical unit. Real-life minutes per day would probably be sufficiently precise and would be easy for players to calculate.

Out of curiosity, why do you say that servers "need" the game to run at 24 minutes per in-game day?

**Reply by James103** [reply]
> Out of curiosity, why do you say that servers "need" the game to run at 24 minutes per in-game day?

The 24 minutes per day was just an arbitrary value that I made up. It could just as easily be 5, 6, 10, 12, 20, 30, or any other value smaller than 1440.

**Reply by kaomoneus** [reply]
I think it might be something like *SetDateWindow*, where you can select how many user's days, hours and minutes the game year should last. What do you think guys?

**Reply by Ufiby** [reply]
In general, everything is fine, the game has a balance in terms of playing time. There is a little confusion with finances. You can make the coefficient x1, x10, x100 and x1000 more modest. It doesn't matter to me that the game is compared to the reality of time. The main thing is that the game with a slowing pace does not depend on real time.

**Reply by kaomoneus** [reply]
Main idea was to keep payback periods same in terms of game time. Let's assume game year lasts one week. Then if you leave game for week, then next week you should observe results as you would leave it for 15 mins in vanilla mode.

If you want money scale to be different, I believe there are already plenty of such settings. But it certainly will break game balance.

...or I missing something in your point?

**Reply by Ufiby** [reply]
I mean, you made a slow-motion game. And it gives conditionally depends on real time. For example, 1 year is equal to 15 minutes, or 1 year is equal to 1 hour, 1 year is equal to 1 day. It's not necessary for me. The main thing is to show finances and be more modest.
![FINANCE](https://user-images.githubusercontent.com/46504018/147780330-3802ecd3-0481-469d-820e-f1333634bc1f.png)
You made this one on top, and at the bottom I suggest that the numbers be modest.
x 100 or more, tested, it's too long. I think not necessarily. So up to a maximum of 100.

**Reply by kaomoneus** [reply]
@Ufiby , please take a look at alternative solution:
https://github.com/kaomoneus/OpenTTD/issues/10

Here's a screenshot (year pace = 1 hour)
![Screen Shot 2022-01-06 at 7 45 32 PM](https://user-images.githubusercontent.com/857042/148412108-f29e11c8-fe35-4130-af96-b4e78dfeaa96.png)

**Reply by 2TallTyler** [reply]
I like your alternative of scaling payments down rather than scaling costs up. Smaller numbers are easier for mental math.

**Reply by Ufiby** [reply]
It's better this way. Perfect. It has become easier to show finances.

**Reply by kaomoneus** [reply]
Hi folks! Made some fixes:
* Fixed time format to be HH:MM
* Fixed pricing, now costs remains same as in vanilla mode, no matter what you game pace is. But if price is non integer though, it's shown with decimal part.
* Synced with mainstream [master](https://github.com/OpenTTD/OpenTTD) branch, namely be352ce1b6f (I hope there are not so many bugs in `master` comparing to release branches)

So, here's new release:
[SlowPace 1.2 Release](https://github.com/kaomoneus/OpenTTD/releases/tag/12.2.slowpace-1.2)

Any tests and remarks are welcome!

![image](https://user-images.githubusercontent.com/857042/183966628-10449551-d795-47da-9932-7e7b68275b21.png)

**Reply by Ufiby** [reply]
This is quite realistic, and it is much easier to calculate the price. I have a question for you, can you change the "DURATION OF the GAME YEAR" to choose any option from 1 to 255, and not like now, for 15 minutes, for hours and days?
![Openttd](https://user-images.githubusercontent.com/46504018/184192229-1bd4756b-896f-4eb4-bc96-4d7b1f137565.png)

I'll thank you anyway for your hard work. I'm looking forward to it being finished and implemented. :)

**Reply by kaomoneus** [reply]
I think we could do it, but option should be clean and intuitive.
1. I propose to introduce *"Custom"* option with pop-up dialog.
2. We need to provide some units for this number. If user will input `123` what that would mean?

**Reply by Ufiby** [reply]
This means that the game will last approximately 25 hours in real time, which is equal to 1 year in the game.

**Reply by kaomoneus** [reply]
I would rather think `123` is 123 hours.

**Reply by Ufiby** [reply]
Hmm, I thought differently, it is calculated by ticks and depends on the factor (user's choice).
![Time](https://user-images.githubusercontent.com/46504018/184407667-84feee9f-ef5a-42be-9619-1a9abc040820.png)

**Reply by kaomoneus** [reply]
Now that makes sense to me as a developer. But we should keep user brains safe from such calculations, haha :-)
So ideally it should very clean metric. As a candidate I would propose to set annual period in real time. It might be implemented as
* single integer field which is "minutes".
* Triple of fields: "DD:HH:MM". So in this case user sets real annual period in days, hours and minutes. For small periods DD and HH just remain zeros.

**Reply by Ufiby** [reply]
I don't know how you calculate it as it is more convenient. How to fix the price and income. If you change it by minutes, what will it mean? I choose 300 minutes, it will be in the game 1 year for 300 minutes in real time right?

**Reply by kaomoneus** [reply]
Yes, like this. I think we can put some short description in pop-up.

**Reply by James103** [reply]
> Yes, like this. I think we can put some short description in pop-up.

Something like the second image, except with a section like in the first image below or above the textbox?
![image](https://user-images.githubusercontent.com/37945304/184472665-cf1ca610-a745-4a04-b329-fd1eafd0b59e.png)
![image](https://user-images.githubusercontent.com/37945304/184472683-f799131a-c1f5-42a9-99b2-1eb817df4428.png)

**Comment by kaomoneus** [comment]
Released new version!

[OpenTTD 12.2 SlowPace 1.3](https://github.com/kaomoneus/OpenTTD/releases/tag/12.2.slowpace-1.3)

* Added game year customization. Among with preset game year durations (one hour, one day, one week and one month) user can select "custom" and set its own duration.
* Game year duration dropdown has been moved from "New Game" dialog into "Intro" dialog.
* Slow pace is now available for scenarios!

MCAAHNY!!

![image](https://user-images.githubusercontent.com/857042/209360463-deacc15b-4f30-4c48-b1ae-c3d687e79141.png)
![image](https://user-images.githubusercontent.com/857042/209360542-d46e4051-e24b-4d31-a2d8-208b5817ac16.png)

**Reply by benjamin-kirkbride** [reply]
This is awesome! Do we expect this to be mainlined at some point? This has always been my #1 issue with this game.

**Comment by kaomoneus** [comment]
New release: synced with 13.4
[13.4.slowpace-1.3.1](https://github.com/kaomoneus/OpenTTD/releases/tag/13.4.slowpace-1.3.1)
